rts/motoko-rts/src/persistence/compatibility.rs

```diff
@@ -0,0 +1,159 @@
+//! Persistent type compatibility check.
+//! Determines whether the a new actor type is compatible with the existing persistent state.
+//! Engages the existing IDL subtype check functionality.
+
+use crate::{
+    barriers::write_with_barrier,
+    bitrel::BitRel,
+    constants::WORD_SIZE,
+    memory::{alloc_blob, Memory},
+    types::{Value, Words},
+};
+
+const DEFAULT_VALUE: Value = Value::from_scalar(0);
+
+pub struct TypeDescriptor {
+    // Blob with candid-encoded type defintions.
+    candid_data: Value,
+    // Blob with a list of `u32` offsets referring to the `candid_data`.
```

Comment:
CRUSSO: GOT HERE

Reply:
I wonder if the descriptor couldn't be simplified to just a single blob (`candid_data`) containing the idl header followed by the index of the main actor type (instead of the usual candid payload). Then you could use a slightly modified parse_idl_header to read out the type_offsets, validating the header and advancing the buffer to the index of the one and only main actor.

The advantage would be less code plus additional validation. But perhaps I'm missing something.

Reply:
Ah, maybe what I am missing is that the other calls to idl_sub in deserialization need access to the main type table and this could be moved by gc?

Reply:
Yes, the deserialization also needs the main table, apart from the upgrade check. The blobs can be moved by the GC, this is also why I can only persist offsets, not the pointers into the table.